components/gitpod-protocol/src/gitpod-service.ts

```diff
@@ -62,6 +62,7 @@ import { InstallationAdminSettings, TelemetryData } from "./installation-admin-p
 import { ListUsageRequest, ListUsageResponse } from "./usage";
 import { SupportedWorkspaceClass } from "./workspace-class";
 import { BillingMode } from "./billing-mode";
+import { CostCenter } from "@gitpod/usage-api/lib/usage/v1/usage.pb";
```

Comment:
Aha, I knew this lined looked suspicious. Werft failed with:

```
[components/gitpod-protocol:lib] src/gitpod-service.ts(65,28): error TS2307: Cannot find module '@gitpod/usage-api/lib/usage/v1/usage.pb' or its corresponding type declarations.
```

@svenefftinge or @easyCZ, do you know how I can import the `CostCenter` type definition from `usage-api` here?

Reply:
You cannot out of the box because it would be cyclic. Redefine (copy) the shape in protocol.ts is the current way to go. We do this with some other shapes already.

Reply:
AFAIK, node doesn't care about cyclic imports. It should be possible to import it, I'll have a look later tonight.

Reply:
It's importable if you add the following to the `gitpod-protocol/BUILD.yaml`

```
deps:
 - components/usage-api/typescript:lib
```
and to `package.json`
```
"@gitpod/usage-api": "0.1.5",
```

Basically mirror the setup that's done in `server` to import it.

Reply:
I would appreciate not introducing cyclic dependencies across packages.  What is the order in which leeway builds them?

Reply:
It's transitive, not cyclic. Usage -> protocol -> server

Reply:
ah, ok. that sounds way better. Thanks for clarifying 🙏

Reply:
Woohoo, many thanks @easyCZ and @svenefftinge for explaining and getting to the bottom of this! 🙌 I learned something new today.

Reply:
FWIW, I really like the idea of not redefining the CostCenter but using the generated TS object. Thanks for doing this.

Reply:
@easyCZ Interesting, I believe I did as you suggested, but now the Werft error changed to:

```
[components/gitpod-protocol:lib] error An unexpected error occurred: "https://registry.yarnpkg.com/@gitpod%2fusage-api: Not found".
```

Reply:
If there's the appropriate definition in BUILD.yaml, leeway should automatically reference the local version in the yarn workspace.

Reply:
Here is what my PR currently has:

<img width="1069" alt="Screenshot 2022-11-09 at 10 21 24" src="https://user-images.githubusercontent.com/599268/200791169-bf025a4d-925b-4a78-a234-42760dd57f99.png">
<img width="1069" alt="Screenshot 2022-11-09 at 10 21 37" src="https://user-images.githubusercontent.com/599268/200791177-4212f17d-d9fe-43c7-9348-4f4a8aaad76c.png">

Is this correct?

Reply:
If I add the `dep` to the `name: lib` / `type: yarn` section (instead of the `name: all` / `type: generic` section as above), I get yet another error:

```
error="dependency cycle found: components:all -> components:all-docker -> components/dashboard:docker -> components/dashboard:app -> components/gitpod-protocol:lib -> components/usage-api/typescript:lib -> components/gitpod-protocol:lib"
```

Maybe there is in fact a cycle if we do this?

Reply:
There exists a cycle, but it's not caused by us wanting to import usage. Here's a [PR](https://github.com/gitpod-io/gitpod/pull/14546) which breaks that cycle by making things more cohesive

Reply:
Awesome, many thanks for fixing the cycle(s)! 🙏 Now that your PR is merged and I've rebased, the build passes again. 🎉

Time for another round of testing & tweaking. 🔍